File: p7-skill/rpc/socket-demo/src/main/java/com/terwergreen/kotlin/ServerDemo.kt
package com.terwergreen.kotlin

import java.io.IOException
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.Executors

/**
 * 服务端
 *
 * @name: ServerDemo
 * @author: terwer
 * @date: 2022-05-09 23:14
 **/
object ServerDemo {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        // 1.创建一个线程池，如果有客户端链接就创建一个线程与之通信
        val executorService = Executors.newCachedThreadPool()
        // 2.创建ServerSocket
        val serverSocket = ServerSocket(9999)
        println("服务器已启动")
        while (true) {
            // 3.监听客户端
            val socket = serverSocket.accept()
            println("有客户端链接")
            executorService.execute { handle(socket) }
        }
    }

    private fun handle(socket: Socket) {
        try {
            println("线程ID:" + Thread.currentThread().id + ",线程名称：" + Thread.currentThread().name)
            // 从连接中取出输入流
            val inputStream = socket.getInputStream()
            val b = ByteArray(1024)
            val read = inputStream.read(b)
            println("客户端发过来的消息：" + String(b, 0, read))

            // 链接中取出输出流并回话
            val outputStream = socket.getOutputStream()
            outputStream.write("没有".toByteArray())
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                socket.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}

File: p7-skill/rpc/socket-demo/src/main/java/com/terwergreen/ruby/server_demo.rb
require 'java'

java_import 'java.util.concurrent.Executors'
java_import 'java.net.ServerSocket'
java_import 'java.net.SocketInputStream'
java_import 'java.lang.Byte'

# JRuby版socket服务端
#
# https://github.com/jruby/jruby/wiki/JRuby-Reference#Arrays
#
# @name: ServerDemo
# @author: terwer
# @date: 2022-05-09 21:04
class ServerDemo
  def self.main()
    executorService = Executors.newCachedThreadPool()
    serverSocket = ServerSocket.new(9999)
    puts("jbury server is running...")
    while (true)
      socket = serverSocket.accept()
      runnable_lambda_fun = -> { self.handle(socket) }
      executorService.execute(runnable_lambda_fun)
    end
  end

  def self.handle(socket)
    begin
      print("receiced socket here=>")
      puts(socket)
      inputStream = socket.getInputStream()
      b = Java::byte[1024].new
      read = inputStream.read(b)
      puts "received msg from client=> #{b}"

      outputStream = socket.getOutputStream()
      out_bytes = "server msg send from jruby".to_java_bytes
      outputStream.write(out_bytes)
    rescue => e
      socket.close()
      puts "Exception Occurred #{e.class}. Message: #{e.message}. Backtrace:  \n #{e.backtrace.join("\n")}"
      Rails.logger.error "Exception Occurred #{e.class}. Message: #{e.message}. Backtrace:  \n #{e.backtrace.join("\n")}"
    end
  end
end

ServerDemo.main()